meilisearch-core/src/settings.rs

```diff
@@ -0,0 +1,181 @@
+            synonyms: settings.synonyms.into(),
+            index_new_fields: settings.index_new_fields.into(),
+        })
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum UpdateState<T> {
+    Update(T),
+    Clear,
+    Nothing,
+}
+
+impl <T> From<Option<Option<T>>> for UpdateState<T> {
+    fn from(opt: Option<Option<T>>) -> UpdateState<T> {
+        match opt {
+            Some(Some(t)) => UpdateState::Update(t),
+            Some(None) => UpdateState::Clear,
+            None => UpdateState::Nothing,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct RankingRuleConversionError;
+
+impl std::fmt::Display for RankingRuleConversionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "impossible to convert into RankingRule")
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum RankingRule {
+    Typo,
+    Words,
+    Proximity,
+    Attribute,
+    WordsPosition,
+    Exact,
+    Asc(String),
+    Dsc(String),
+}
+
+impl ToString for RankingRule {
+    fn to_string(&self) -> String {
+        match self {
+            RankingRule::Typo => "_typo".to_string(),
+            RankingRule::Words => "_words".to_string(),
+            RankingRule::Proximity => "_proximity".to_string(),
+            RankingRule::Attribute => "_attribute".to_string(),
+            RankingRule::WordsPosition => "_words_position".to_string(),
+            RankingRule::Exact => "_exact".to_string(),
+            RankingRule::Asc(field) => format!("asc({})", field),
+            RankingRule::Dsc(field) => format!("dsc({})", field),
+        }
+    }
+}
+
+impl FromStr for RankingRule {
+    type Err = RankingRuleConversionError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let rule = match s {
+            "_typo" => RankingRule::Typo,
+            "_words" => RankingRule::Words,
+            "_proximity" => RankingRule::Proximity,
+            "_attribute" => RankingRule::Attribute,
+            "_words_position" => RankingRule::WordsPosition,
+            "_exact" => RankingRule::Exact,
+            _ => {
+                let captures = RANKING_RULE_REGEX.captures(s).ok_or(RankingRuleConversionError)?;
+                match (captures.get(1).map(|m| m.as_str()), captures.get(2)) {
+                    (Some("asc"), Some(field)) => RankingRule::Asc(field.as_str().to_string()),
+                    (Some("dsc"), Some(field)) => RankingRule::Dsc(field.as_str().to_string()),
+                    _ => return Err(RankingRuleConversionError)
+                }
+            }
+        };
+        Ok(rule)
+    }
+}
+
+impl RankingRule {
+    pub fn field(&self) -> Option<&str> {
+        match self {
+            RankingRule::Asc(field) | RankingRule::Dsc(field) => Some(field),
+            _ => None,
+        }
+    }
+
+    pub fn from_vec(rules: Vec<&str>) -> Result<Vec<RankingRule>, RankingRuleConversionError> {
```

Comment:
Please [do not force an allocation](https://github.com/meilisearch/MeiliSearch/pull/441/files#diff-af49e5999ad6279210f4e496c00a84caR46) when you don't need one and you will not reuse it.
Change this method to accept an `Iterator` of `AsRef<str>` instead.